[CSS] Implement @scope implicit

mdubet opened the pull request:
#### cee867893141fabf4fdeed3603a8986b74af3f88
<pre>
[CSS] Implement @scope implicit
<a href="https://bugs.webkit.org/show_bug.cgi?id=266399">https://bugs.webkit.org/show_bug.cgi?id=266399</a>
<a href="https://rdar.apple.com/119659940">rdar://119659940</a>

Reviewed by Antti Koivisto.

When the @scope rule doesn&apos;t have a &lt;scope-start&gt; prelude,
the scoping root is the parent element of the owner node of the stylesheet
where the @scope rule is defined.

We store in the @scope rule a pointer to the owner StyleSheetContent,
which will allow us to find the parent element of it at rule matching time,
to determine the scoping root(s).

<a href="https://drafts.csswg.org/css-cascade-6/#scope-limits">https://drafts.csswg.org/css-cascade-6/#scope-limits</a>

* LayoutTests/fast/css/scope-at-rule-expected.html:
* LayoutTests/fast/css/scope-at-rule-green-14.css: Added.
(@scope):
* LayoutTests/fast/css/scope-at-rule.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/at-scope-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/at-scope-parsing.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-implicit-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-implicit-external-expected.txt:
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRuleScope::styleSheetContents const):
(WebCore::StyleRuleScope::setStyleSheetContents):
* Source/WebCore/css/StyleRule.h:
* Source/WebCore/css/StyleSheetContents.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeScopeRule):
* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::ruleMatches):
(WebCore::Style::ElementRuleCollector::scopeRulesMatch):
* Source/WebCore/style/ElementRuleCollector.h:
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addChildRule):

Canonical link: <a href="https://commits.webkit.org/272390@main">https://commits.webkit.org/272390@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b43d30c118695b6127d60347b6c7ccefb67def53

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/31492 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/10167 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/33196 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/33980 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/28522 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/32262 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/12516 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/7414 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/28138 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/31832 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/8561 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/28104 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/7367 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/7530 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/28013 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/35324 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/28623 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/28459 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/33663 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/7606 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/5624 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/31507 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9263 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/27818 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7398 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8295 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8111 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/style/StyleScopeRuleSets.cpp:
```diff
@@ -234,7 +234,11 @@ void ScopeRuleSets::appendAuthorStyleSheets(const Vector<RefPtr<CSSStyleSheet>>&
+        const Node* ownerNode = cssSheet->ownerNode();
+        if (ownerNode)
+            ownerNode = ownerNode->parentNode();
+        const ContainerNode* container = dynamicDowncast<ContainerNode>(ownerNode);
```
```suggestion
        auto* container = dynamicDowncast<ContainerNode>(ownerNode);
```

mdubet replied:
Why lose the const-qualifier ? `const auto*` ?

nt1m replied:
Sure, you should probably use some type of smart pointer for nodes btw

mdubet replied:
Ok so not `auto*` then 😅

anttijk commented on Source/WebCore/style/RuleSet.h:
```diff
@@ -119,8 +119,15 @@ class RuleSet : public RefCounted<RuleSet> {
+        Ref<const StyleRuleScope> scopeRule;
+        ScopeRuleIdentifier parent;
+        RefPtr<const ContainerNode> ownerNode;
+    };
```
I know I suggested this approach but I now think there is actually a problem with putting Node pointers into RuleSets. The issue is that we can share Resolvers (and so RuleSets) between different shadow trees (`Scope::createOrFindSharedShadowTreeResolver()`). While this sharing could be disallowed in this case it would probably be architecturally better to not have Nodes here.

The alternative strategy of adding StyleSheetContents pointer to StyleRuleScope and find the owner node(s) via it (and either disabling caching in `StyleSheetContents::isCacheable()` or iterating over clients to find the relevant owners) would probably be cleaner and less code.

anttijk commented on Source/WebCore/style/ElementRuleCollector.cpp:
```diff
@@ -677,10 +680,34 @@ std::pair<bool, std::optional<Vector<ElementRuleCollector::ScopingRootWithDistan
+            auto appendNode= [&](const auto* client) {
+                const auto* owner = client->ownerNode();
+                if (owner)
+                    owner = owner->parentNode();
```
There could be a separate variable not called `owner` for the implicit scope parent, like say `implicitScopeParent`.

mdubet replied:
I forgot to update the name of the variable in the lookup after .. https://github.com/WebKit/WebKit/pull/25918

anttijk commented on Source/WebCore/style/ElementRuleCollector.cpp:
```diff
@@ -677,10 +680,34 @@ std::pair<bool, std::optional<Vector<ElementRuleCollector::ScopingRootWithDistan
-            // https://bugs.webkit.org/show_bug.cgi?id=266399
-            return false;
+            // The scoping root is the parent element of the owner node of the stylesheet where the @scope rule is defined. (If no such element exists, then the scoping root is the root of the containing node tree.
+            RefPtr<const StyleSheetContents> styleSheetContents = rule->styleSheetContents();
```
You can just say `RefPtr styleSheetContents =` I think.

nt1m replied:
```suggestion
            RefPtr styleSheetContents = rule->styleSheetContents();
```

anttijk commented on Source/WebCore/style/ElementRuleCollector.cpp:
```diff
@@ -677,10 +680,34 @@ std::pair<bool, std::optional<Vector<ElementRuleCollector::ScopingRootWithDistan
+
+            // For @import rule, we need to find the enclosing parent.
+            while (const auto* owner = styleSheetContents->ownerRule())
+                styleSheetContents = owner->parentStyleSheet();
```
There is `StyleSheetContents::rootStyleSheet()` that already does this.

anttijk commented on Source/WebCore/style/ElementRuleCollector.cpp:
```diff
@@ -677,10 +680,34 @@ std::pair<bool, std::optional<Vector<ElementRuleCollector::ScopingRootWithDistan
+                return false;
+
+            auto appendNode= [&](const auto* client) {
+                const auto* owner = client->ownerNode();
```
You could check the the node is actually in the current document (by comparing document() ptr) and ignore those that aren't.

nt1m commented on Source/WebCore/style/ElementRuleCollector.cpp:
```diff
@@ -677,10 +680,34 @@ std::pair<bool, std::optional<Vector<ElementRuleCollector::ScopingRootWithDistan
+            if (!styleSheetContents)
+                return false;
+
+            auto appendNode= [&](const auto* client) {
```
```suggestion
            auto appendNode = [&](const auto* client) {
```

anttijk replied:
appendImplicitScopeParentNode or something similarly descriptive?

anttijk commented on Source/WebCore/css/StyleRule.h:
```diff
@@ -406,6 +409,8 @@ class StyleRuleScope final : public StyleRuleGroup {
     CSSSelectorList m_originalScopeStart;
     CSSSelectorList m_originalScopeEnd;
+    // Pointer to the owner StyleSheetContent to find the implicit scope (when there is no <scope-start>)
+    const StyleSheetContents* m_styleSheetOwner { nullptr };
```
I don't see you clearing this anywhere, you can probably leave a dangling pointer behind. This should really be a smart pointer (WeakPtr probably, StyleSheetContents should be CanMakeWeakPtr).

mdubet replied:
Fixed ✅

anttijk commented on Source/WebCore/css/StyleRule.h:
```diff
@@ -406,6 +409,8 @@ class StyleRuleScope final : public StyleRuleGroup {
     CSSSelectorList m_originalScopeStart;
     CSSSelectorList m_originalScopeEnd;
+    // Pointer to the owner StyleSheetContent to find the implicit scope (when there is no <scope-start>)
+    WeakPtr<const StyleSheetContents> m_styleSheetOwner { nullptr };
```
I think it might be clearer to just set this unconditionally and check wether use it or not during selector matching.

mdubet committed:
[CSS] Implement @scope implicit
https://bugs.webkit.org/show_bug.cgi?id=266399
rdar://119659940

Reviewed by Antti Koivisto.

When the @scope rule doesn't have a <scope-start> prelude,
the scoping root is the parent element of the owner node of the stylesheet
where the @scope rule is defined.

We store in the @scope rule a pointer to the owner StyleSheetContent,
which will allow us to find the parent element of it at rule matching time,
to determine the scoping root(s).

https://drafts.csswg.org/css-cascade-6/#scope-limits

* LayoutTests/fast/css/scope-at-rule-expected.html:
* LayoutTests/fast/css/scope-at-rule-green-14.css: Added.
(@scope):
* LayoutTests/fast/css/scope-at-rule.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/at-scope-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/at-scope-parsing.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-implicit-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-implicit-external-expected.txt:
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRuleScope::styleSheetContents const):
(WebCore::StyleRuleScope::setStyleSheetContents):
* Source/WebCore/css/StyleRule.h:
* Source/WebCore/css/StyleSheetContents.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeScopeRule):
* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::ruleMatches):
(WebCore::Style::ElementRuleCollector::scopeRulesMatch):
* Source/WebCore/style/ElementRuleCollector.h:
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addChildRule):

Canonical link: https://commits.webkit.org/272390@main
